packages/flutter_tools/lib/src/windows/build_windows.dart

```diff
@@ -43,74 +43,105 @@ Future<void> buildWindows(WindowsProject windowsProject, BuildInfo buildInfo, {
   }
 
   // Ensure that necessary emphemeral files are generated and up to date.
-  _writeGeneratedFlutterProperties(windowsProject, buildInfo, target);
-  createPluginSymlinks(windowsProject.project);
+  _writeGeneratedFlutterConfig(windowsProject, buildInfo, target);
+  createPluginSymlinks(windowsProject.parent);
 
   final VisualStudio visualStudio = visualStudioOverride ?? VisualStudio(
     fileSystem: globals.fs,
     platform: globals.platform,
     logger: globals.logger,
     processManager: globals.processManager,
   );
-  final String vcvarsScript = visualStudio.vcvarsPath;
-  if (vcvarsScript == null) {
+  final String cmakePath = visualStudio.cmakePath;
+  if (cmakePath == null) {
     throwToolExit('Unable to find suitable Visual Studio toolchain. '
         'Please run `flutter doctor` for more details.');
   }
 
-  final String buildScript = globals.fs.path.join(
-    Cache.flutterRoot,
-    'packages',
-    'flutter_tools',
-    'bin',
-    'vs_build.bat',
-  );
-
-  final String configuration = toTitleCase(getNameForBuildMode(buildInfo.mode ?? BuildMode.release));
-  final String solutionPath = windowsProject.solutionFile.path;
-  final Stopwatch sw = Stopwatch()..start();
+  final String buildModeName = getNameForBuildMode(buildInfo.mode ?? BuildMode.release);
+  final Directory buildDirectory = globals.fs.directory(getWindowsBuildDirectory());
   final Status status = globals.logger.startProgress(
     'Building Windows application...',
     timeout: null,
   );
+  try {
+    await _runCmakeGeneration(cmakePath, buildDirectory, windowsProject.cmakeFile.parent);
+    await _runBuild(cmakePath, buildDirectory, buildModeName);
+  } finally {
+    status.cancel();
+  }
+}
+
+Future<void> _runCmakeGeneration(String cmakePath, Directory buildDir, Directory sourceDir) async {
+  final Stopwatch sw = Stopwatch()..start();
+
+  await buildDir.create(recursive: true);
   int result;
   try {
-    // Run the script with a relative path to the project using the enclosing
-    // directory as the workingDirectory, to avoid hitting the limit on command
-    // lengths in batch scripts if the absolute path to the project is long.
     result = await processUtils.stream(
       <String>[
-        buildScript,
-        vcvarsScript,
-        globals.fs.path.basename(solutionPath),
-        configuration,
+        cmakePath,
+        '-S',
+        sourceDir.path,
+        '-B',
+        buildDir.path,
+        '-G',
+        'Visual Studio 16 2019',
```

Comment:
nit: maybe put this in a constant, with some docs on where exactly it comes from, like "we only support this particular sdk, see doctor blah blah".

Reply:
Done.

It hadn't occurred to me that this may mean that the next major version of VS won't actually work since the generator is hard-coded; I don't know if these values are forward-compatible. But since this is in the tool, not the ecosystem of plugins and projects, we can easily add switching logic here later if that turns out to be the case. (Unlike the requirement in the old method of having a hard-coded toolchain version in the project file, which was one of the things that really worried me about the ecosystem longer term.) I added a comment to that effect on the constant.